### PR TITLE
Added support for app open ads

### LINKED ISSUE #23: Add new ad format `App Open Ads`
More information: [https://developers.google.com/admob/android/app-open](https://developers.google.com/admob/android/app-open)

### PR DESCRIPTION
**Pull request by britzl**
Fixes #23

### CODE REVIEW

**Comment by britzl** [comment]
The problem with the current solution is that it doesn't hook in before the DefoldActivity is shown the first time. Once the app is running and put in the background and then back to the foreground it works though.

In order to also show an add on app start we also need to extend the Application class (https://developers.google.com/admob/android/app-open#extend)

**Comment by AGulev** [comment]
> In order to also show an add on app start we also need to extend the Application class (https://developers.google.com/admob/android/app-open#extend)

I think it's possible to avoid:

```
static bool g_IsAppInitialized = false;

static void OnCreateApp(JNIEnv* env, jobject activity)
{
  // just to make sure it's ccalled only once
  // there is an issue with UnregisterOnActivityCreateListener
  if (!g_IsAppInitialized)
  {
    // Don't usee dmAndroid::LoadClass on OnCreate on some android versions g_AndroidApp->activity->clazz
    // isn't initialized yet!
    jclass na_class = env->FindClass("android/app/NativeActivity");
    jmethodID get_class_loader = env->GetMethodID(na_class,"getClassLoader", "()Ljava/lang/ClassLoader;");
    jobject cls = env->CallObjectMethod(activity, get_class_loader);
    jclass class_loader = env->FindClass("java/lang/ClassLoader");
    jmethodID find_class = env->GetMethodID(class_loader, "loadClass", "(Ljava/lang/String;)Ljava/lang/Class;");
    jstring str_class_name = env->NewStringUTF("com/defold/admob/AdmobJNI");
    jclass admob_cls = (jclass)env->CallObjectMethod(cls, find_class, str_class_name);
    env->DeleteLocalRef(str_class_name);
    jmethodID method = env->GetStaticMethodID(admob_cls, "AdMob_OnCreate", "(Landroid/app/Activity;)V");
    env->CallStaticVoidMethod(admob_cls, method, activity);
    g_IsAppInitialized = true;
  }
}

struct AdMobStaticInit
{
  AdMobStaticInit()
  {
    dmAndroid::RegisterOnActivityCreateListener(OnCreateApp);
  }
} g_AdMobStaticInit;
```

```java
public static void AdMob_OnCreate(final Activity appActivity) {
// your OnCreate logic here
}
```

It should work

File: extension-admob/ext.manifest [hunk 1]
```diff
@@ -3,7 +3,7 @@ name: "AdMobExt"
 platforms:
     android:
         context:
-            aaptExtraPackages: ["androidx.work", "com.google.android.gms.common", "androidx.startup"]
+            aaptExtraPackages: ["androidx.work", "com.google.android.gms.common", "androidx.startup", "androidx.lifecycle"]
```
Comment (AGulev) [inline extension-admob/ext.manifest:6]: Do we still need these? I though we generate them now.

File: extension-admob/src/admob_android.cpp [hunk 2]
```diff
@@ -114,33 +119,53 @@ static void InitJNIMethods(JNIEnv* env, jclass cls)
...
 }
 
 void Finalize_Ext()
 {
 }
 
-void Initialize(const char* defoldUserAgent)
+void Initialize()
+{
+    CallVoidMethodChar(g_admob.m_AdmobJNI, g_admob.m_Initialize);
```
Comment (AGulev) [inline extension-admob/src/admob_android.cpp:153]: /extension-admob/src/admob_android.cpp
	Line 153: no matching function for call to 'CallVoidMethodChar'
    CallVoidMethodChar(g_admob.m_AdmobJNI, g_admob.m_Initialize);
    ^~~~~~~~~~~~~~~~~~

Comment (AGulev) [reply extension-admob/src/admob_android.cpp:153]: did you mean `CallVoidMethod()`?

**Review by AGulev — CHANGES_REQUESTED** [review]
Android can't be compiled, it seems like wrong function used.

Should it be shown on iOS right after I run app?

File: extension-admob/src/admob_android.cpp (hunk 2 above)
Comment (britzl) [reply extension-admob/src/admob_android.cpp]: Sigh, never do last minute changes without testing them....

File: extension-admob/ext.manifest (hunk 1 above)
Comment (britzl) [reply extension-admob/ext.manifest:6]: Hmm, let me try without them.

**Comment by britzl** [comment]
> Should it be shown on iOS right after I run app? Because now it's shown only when I lock unclock phone

Let me have a look!

File: extension-admob/ext.manifest (hunk 1 above)
Comment (britzl) [reply extension-admob/ext.manifest]: Yep, no need anymore!

**Comment by AGulev** [comment]
Great!

**Review by AGulev — APPROVED** [review]
It is great addition!!!

**Pull request merged**